### ISSUE TITLE
Pre-Built Windows x64 .exe's Crash on Machines without nvcuda.dll

### ISSUE THREAD

**Issue by mchildsCO76** [cdcseacave/openMVS#786]
The currently pre-build Windows x64 .exe's for OpenMVS 2.0 run great on machines that have CUDA (nVidia) support. However, if you try and run DensifyPointCloud.exe or RefineMesh.exe on a machine that does not support CUDA, even if you supply the option flags to not use CUDA, the .exe immediately crashes as nvcuda.dll cannot be found on the machine.

**To Reproduce**
Run the prebuild OpenMVS 2.0 DensifyPointCloud.exe from the command line (no parameters) on a non-CUDA machine

**Expected behavior**
The crash should only happen if the flag to use CUDA is provided.

**Suggestion to Fix**
Simply marking the nvcuda.dll as Delay-Load in the Visual Studio linker settings should prevent this. Then the nvcuda.dll won't be loaded until a call to any functions in that DLL is made. So long as the 'use CPU' flag is set, no CUDA function should ever be called and everything should be fine. This setting is in Visual Studio under the project Configuration Properties -> Linker -> Input -> Delay Loaded DLLs. Just adding nvcuda.dll there *should* fix the issue.

I will likely just build OpenMVS myself and add the delay-load option so that 2 separate .exe's aren't needed for CUDA and non-CUDA machines. If it was in the pre-build DLLs then it would just work for everyone out of the box regardless of machine configuration.

**Comment by mchildsCO76** [comment]
It seems a delay-load is not enough as CUDA functions are called in DensifyPointCloud.exe even if '--cuda-device -1' is provided to enable CPU only mode. It seems that _initDevice_ in UtilCUDA.cpp is called no matter what at startup, at which point nvcuda.dll would be loaded. I will try making a nvcuda.dll stub to use on machines without CUDA to see if that can work around the issue.

**Comment by cdcseacave** [comment]
delay loading nvcuda.dll is a great idea, pls add it if you know how; no cuda call is needed if --cuda-device is -1, maybe I simply forgot to condition the initDevice call

**Comment by mchildsCO76** [comment]
The code that initializes the list of CUDA devices is still called even if --cuda-device -1 is provided, even though the list isn't used.

I've attached the nvcuda_stub Visual Studio 2019 project that I created which stubs out the required functions in case anyone else wants to use it. Now it works fine on non-CUDA enabled machines. To use it, you just rename nvcuda_stub.dll to nvcuda.dll on such machines and place it in the path of the .exe's. Or, when you run the process, first add the path to wherever the nvcuda.dll to the PATH environment variable.

If you do update the code so that CUDA isn't called at all if --cuda-device -1 is provided, then the delay load would be sufficient. While I know how to add that in Visual Studio itself, I'm not sure where to stick that in the CMake options.

[nvcuda_stub.zip](https://github.com/cdcseacave/openMVS/files/8311753/nvcuda_stub.zip)

**Comment by cdcseacave** [comment]
pls try the latest develop, it should not call CUDA functions if  --cuda-device -1  set

**Comment by mchildsCO76** [comment]
I'll try pulling and building development when I get a chance. The stub DLL solution is working well for me now even with the pre-build DLLs, but with the changes I could ditch that in the future.

**Comment by mchildsCO76** [comment]
I still get CUDA calls in DensifyPointCloud.exe with --cuda-device -1 provided as a parameter. Here is the call stack where the first one occurs:

`>	DensifyPointCloud.exe!SEACAVE::CUDA::initDevice(int deviceID) Line 166	C++
 	DensifyPointCloud.exe!MVS::Mesh::InitKernels(int device) Line 4104	C++
 	DensifyPointCloud.exe!MVS::Mesh::Mesh() Line 137	C++
 	DensifyPointCloud.exe!MVS::Scene::Scene(unsigned int _nMaxThreads) Line 67	C++
 	DensifyPointCloud.exe!main(int argc, const char * * argv) Line 255	C++
`

The value of deviceID is -1 (passed in from CUDA::desiredDeviceID which comes from the parameter). It looks like the initDevice comments specify that -1 selects the best available CUDA device, while the --cuda-device command line parameter it comes from indicates to use the CPU for -1. Perhaps the MVS::Mesh constructor needs to not call InitKernels if CUDA::desiredDeviceID is -1 or something.

**Comment by cdcseacave** [comment]
true, thx, I fixed both issues (hopefully) in the latest commit, now -1 means the best GPU, and -2 CPU

**Comment by mchildsCO76** [comment]
I can confirm that the CUDA calls are no longer made when --cuda-device -2 is provided in the latest code.

I also see that --cuda-device has been normalized across all of the .exe's (and added to TextureMesh and ReconstructMesh). Looking forward to those changes in a release!
